### PR TITLE
feat: added rtl direction to ConfigProvider and some components

### PR DESCRIPTION
**Pull request by saeedrahimi**
### 🤔 This is a ...

- [x] New feature
- [x] Bug fix
- [x] Site / document update
- [x] Component style update
- [ ] TypeScript definition update
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
[#4051](https://github.com/ant-design/ant-design/issues/4051)

### 💡 Background and solution
As requirement of this [issue](https://github.com/ant-design/ant-design/issues/4051) components of ant-design doesn't support right to left direction, and those components are not useful in some languages like Persian(Farsi) or Arabic.

Solution: 
I started to develop Rtl style on ant-design and as @zombieJ  mentioned in issue i used ConfigProvider to set direction of component.

layoutDirection property added to ConfigProvider.
```
<ConfigProvider layoutDirection="rtl">
      <Cascader  options={options} placeholder="select an item"/>
</ConfigProvider>
```
in style files added rtl based styles to change direction of displayed items
```
&-picker-small &-picker-clear,
  &-picker-small &-picker-arrow {
    right: @control-padding-horizontal-sm;
    [dir='rtl'] &, .rtl & {
      right: auto;
      left: @control-padding-horizontal;
    }
  }
```
so when one of parent elements has dir=['rtl'] (provided by ConfigProvider) paddings , margins, floats and other properties flipped to display correctly in right-to-left directions.

cascader

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   Added Rtl direction on cascader        |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----
[View rendered components/cascader/demo/right-to-left.md](https://github.com/saeedrahimi/ant-design/blob/cascader-rtl/components/cascader/demo/right-to-left.md)

### CODE REVIEW

**Comment by saeedrahimi** [comment]
Looks snapshots issue resolved in #18036 and it's on master branch. but my changes are on feature branch.
what can i do to fix this pull request?

**Comment by zombieJ** [comment]
We will release minor version this week. which will sync  branch soon.

**Comment by zombieJ** [comment]
How about directly `direction` with value `ltr` & `rtl` like CSS: https://developer.mozilla.org/zh-CN/docs/Web/CSS/direction

@saeedrahimi  @ztplz @afc163

**Comment by saeedrahimi** [comment]
I'm already using it in [these changes ](https://github.com/ant-design/ant-design/blob/ab8b12376756908fd48207f33b96fdc7b58520ea/components/config-provider/index.tsx#L95)
but the problem is some components needs to know the direction of page/view/parent to change behavior. [example](https://github.com/ant-design/ant-design/blob/ab8b12376756908fd48207f33b96fdc7b58520ea/components/cascader/index.tsx#L576) 

I'm not a React developer and i don't know about it's features. but in angular with [bidi](https://material.angular.io/cdk/bidi/api) you can detect inherited direction (RTL/LTR) of a component. i couldn't find anything like bidi in React, so i added variable to ConfigProvider to detect the direction.
I've already added RTL styles to most of components of [ng-zorro](https://ng.ant.design) library, but [they told me](https://github.com/NG-ZORRO/ng-zorro-antd/issues/1762#issuecomment-516315930) it should be implemented in ant-design first.

**Comment by zombieJ** [comment]
I mean change prop name from `layoutDirection` to `direction`.

**Comment by wendzhue** [comment]
兴奋地搓手手

**Comment by saeedrahimi** [comment]
any progress?

File: components/cascader/demo/right-to-left.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  fa-IR: چپ به راست
+---
+
+## zh-CN
+
+省市区级联。
+
+## en-US
+
+Cascader component supports right to left (rtl) direction if placed inside `ConfigProvider` with `direction='rtl'`
```
Comment (zombieJ) [inline components/cascader/demo/right-to-left.md]: This should also doc in `ConfigProvider` page.

Comment (saeedrahimi) [reply components/cascader/demo/right-to-left.md]: `ConfigProvider` documentations updated.

**Comment by saeedrahimi** [comment]
@zombieJ 
`FAIL  components/steps/__tests__/demo.test.js`
it has CI errors in step components.

File: components/config-provider/index.en-US.md [hunk 2]
```diff
@@ -43,3 +43,4 @@ Some component use dynamic style to support wave effect. You can config `csp` pr
 | getPopupContainer | to set the container of the popup element. The default is to create a `div` element in `body`. | Function(triggerNode) | `() => document.body` | 3.11.0 |
 | locale | language package setting, you can find the packages in [antd/es/locale](http://unpkg.com/antd/es/locale/) | object | 3.21.0 |
 | prefixCls | set prefix class | string | ant | 3.12.0 |
+| direction | to set direction of all elements underneath itself, set `rtl` to view underneath components right-to-left styled | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
```
Comment (zombieJ) [inline components/config-provider/index.en-US.md]: * Move Cascader demo into ConfigProvider.
* `set direction of text. See [demo](#your demo link)`

File: components/config-provider/index.zh-CN.md [hunk 3]
```diff
@@ -44,3 +44,4 @@ return (
 | getPopupContainer | 弹出框（Select, Tooltip, Menu 等等）渲染父节点，默认渲染到 body 上。 | Function(triggerNode) | () => document.body | 3.11.0 |
 | locale | 语言包配置，语言包可到 [antd/es/locale](http://unpkg.com/antd/es/locale/) 目录下寻找 | object | - | 3.21.0 |
 | prefixCls | 设置统一样式前缀 | string | ant | 3.12.0 |
+| direction | 要设置自身下方所有元素的方向，请将 `rtl` 设置为从右到左的样式下方查看 | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
```
Comment (zombieJ) [inline components/config-provider/index.zh-CN.md]: `设置文本展示方向。[示例](...)`

File: components/config-provider/index.en-US.md (hunk 2 above)
Comment (zombieJ) [reply components/config-provider/index.en-US.md]: We can add more component demo here.

Comment (saeedrahimi) [reply components/config-provider/index.en-US.md]: all done.

File: components/config-provider/index.zh-CN.md [hunk 4]
```diff
@@ -44,4 +44,4 @@ return (
 | getPopupContainer | 弹出框（Select, Tooltip, Menu 等等）渲染父节点，默认渲染到 body 上。 | Function(triggerNode) | () => document.body | 3.11.0 |
 | locale | 语言包配置，语言包可到 [antd/es/locale](http://unpkg.com/antd/es/locale/) 目录下寻找 | object | - | 3.21.0 |
 | prefixCls | 设置统一样式前缀 | string | ant | 3.12.0 |
-| direction | 要设置自身下方所有元素的方向，请将 `rtl` 设置为从右到左的样式下方查看 | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
+| direction | 设置文本展示方向。[示例](/components/config-provider/#components-config-provider-demo-direction) | string: 'ltr', 'rtl', 'auto' | ltr | 3.21.5 |
```
Comment (zombieJ) [inline components/config-provider/index.zh-CN.md]: `#components-config-provider-demo-direction` is good enought : )

Comment (zombieJ) [reply components/config-provider/index.zh-CN.md]: version: `3.21.5` -> `3.22.0`

**Comment by saeedrahimi** [comment]
@zombieJ could you help me on resolving merge issue. it's about code coverage.

File: components/config-provider/demo/direction.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+          />
+          &nbsp;&nbsp;&nbsp;&nbsp; With search:
+          <Cascader
+            suffixIcon={<Icon type="search" />}
+            options={cascaderOptions}
+            onChange={this.onCascaderChange}
+            placeholder="Select an item"
+            popupPlacement={this.props.popupPlacement}
+            showSearch={this.cascaderFilter}
+          />
```
Comment (zombieJ) [inline components/config-provider/demo/direction.md]: Could we provides more demo instead of Cascader only?

Comment (saeedrahimi) [reply components/config-provider/demo/direction.md]: yes, but for now i just changed styles only for Cascader and each component needs their rtl styles.
I've planned to change each components in separate pull request but I can change more components in this pull request if you want. :)

**Review by afc163 — DISMISSED** [review]

File: components/config-provider/demo/direction.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  };
+
+  cascaderFilter = (inputValue, path) => {
+    return path.some(option => option.label.toLowerCase().indexOf(inputValue.toLowerCase()) > -1);
+  };
+
+  render() {
+    return (
+      <div className="direction-components">
+        <div className="example">
```
Comment (afc163) [inline components/config-provider/demo/direction.md]: Duplicated div for demo wrapper.

File: components/config-provider/demo/direction.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+class App extends React.Component {
+  constructor() {
+    super();
+    this.state = {
+      direction: 'ltr',
+      popupPlacement: 'bottomLeft',
+    };
+  }
```
Comment (afc163) [inline components/config-provider/demo/direction.md]:
```suggestion
state = {
  direction: 'ltr',
  popupPlacement: 'bottomLeft',
};
```

File: components/config-provider/index.tsx [hunk 8]
```diff
@@ -91,11 +96,13 @@ class ConfigProvider extends React.Component<ConfigProviderProps> {
     }
 
     return (
-      <ConfigContext.Provider value={config}>
-        <LocaleProvider locale={locale || legacyLocale} _ANT_MARK__={ANT_MARK}>
-          {children}
-        </LocaleProvider>
-      </ConfigContext.Provider>
+      <div dir={direction}>
```
Comment (afc163) [inline components/config-provider/index.tsx]: Should we add extra div to ConfigProvider?

File: site/theme/template/Content/ComponentDoc.jsx [hunk 9]
```diff
@@ -52,7 +52,13 @@ class ComponentDoc extends React.Component {
...
 
   render() {
-    const { doc, location, intl: { locale }, utils, demos } = this.props;
+    const {
+      doc,
+      location,
+      intl: { locale },
+      utils,
+      demos,
+    } = this.props;
```
Comment (afc163) [inline site/theme/template/Content/ComponentDoc.jsx]: Mixed some unrelated changes

Comment (saeedrahimi) [reply site/theme/template/Content/ComponentDoc.jsx]: I can't exclude these files.
I ran this command and fetched upstream file.
`git checkout upstream/feature -- site/theme/template/Content/Demo.jsx`
but after commit and push files won't disappear from changes.

File: components/config-provider/demo/direction.md (hunk 6 above)
Comment (saeedrahimi) [reply components/config-provider/demo/direction.md]: resolved in commits

File: components/config-provider/index.tsx (hunk 8 above)
Comment (saeedrahimi) [reply components/config-provider/index.tsx]: I couldn't find any proper element to add `dir` attribute to it, so i added this div.
Do you have any suggestion.

File: site/theme/template/Content/ComponentDoc.jsx (hunk 9 above)
Comment (saeedrahimi) [reply site/theme/template/Content/ComponentDoc.jsx]: Just fixed issue by editing file from browser and Github web. :|
I think on my local system it auto-format the file after commit and re-arranges the render method parameters.

File: components/config-provider/index.tsx (hunk 8 above)
Comment (saeedrahimi) [reply components/config-provider/index.tsx]: And something i forgot to say is that i use `dir` attribute in less files to change style of component based on their direction.

Comment (afc163) [reply components/config-provider/index.tsx]: Try pass `dir` in React context to every component.

Comment (saeedrahimi) [reply components/config-provider/index.tsx]: screenshots

or think about a form, if it wrapped with `<div dir="rtl" />` all components get `dir` value from context except those are not React component.
screenshots2

Comment (saeedrahimi) [reply components/config-provider/index.tsx]: another approach is like your suggestion just use `dir` in components and let the user to design their forms or group of elements with rtl direction.

Comment (saeedrahimi) [reply components/config-provider/index.tsx]: like [this pen ](https://codepen.io/saeedr/pen/YzKWLQL) i've just added `dir="rtl` ro `Form` element and everything is ok.

Comment (saeedrahimi) [reply components/config-provider/index.tsx]: Removed extra div on `ConfigProvider`.

**Comment by saeedrahimi** [comment]
@afc163
@zombieJ
any updates?

**Comment by zombieJ** [comment]
hi @saeedrahimi, need more component in the demo of ConfigProvider:
> We can add more component demo here.

File: docs/react/i18n.zh-CN.md [hunk 10]
```diff
@@ -65,7 +65,6 @@ return (
 | 泰米尔语         | ta_IN  |
 | 泰语             | th_TH  |
 | 土耳其语         | tr_TR  |
-| 罗马尼亚语       | ro_RO  |
```
Comment (afc163) [inline docs/react/i18n.zh-CN.md]: Why remove this?

File: components/style/core/rtl.less [hunk 11]
```diff
@@ -0,0 +1,8 @@
+/* stylelint-disable at-rule-no-unknown */
+
+.rtl {
+  direction: rtl;
+}
+.ltr {
+  direction: ltr;
+}
```
Comment (afc163) [inline components/style/core/rtl.less]: Do we need global styles now?

**Comment by saeedrahimi** [comment]
> hi @saeedrahimi, need more component in the demo of ConfigProvider:
> 
> > We can add more component demo here.

I've edited more components and make them support RTL direction and updated ConfigProvider demo with more rtl components.

File: components/style/core/rtl.less (hunk 11 above)
Comment (saeedrahimi) [reply components/style/core/rtl.less]: Maybe not, i'll remove it.

File: components/style/core/rtl.less [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+.rtl .ant-col {
+  float: right;
+}
```
Comment (saeedrahimi) [inline components/style/core/rtl.less]: > Do we need global styles now?

These styles will be needed on some cases like direction of `.ant-col` that i'm used.
so every col wrapped in `.rtl` will be displayed correctly. (Used in ConfigProvider RTL demo)

File: components/style/core/rtl.less (hunk 11 above)
Comment (saeedrahimi) [reply components/style/core/rtl.less]: These styles will be needed on some cases like direction of `.ant-col` that i'm used.
so every col wrapped in `.rtl` will be displayed correctly. (Used in ConfigProvider RTL demo)

**Comment by zombieJ** [comment]
This can be adjust:
屏幕快照 2019-09-25 上午10 41 01

**Comment by zombieJ** [comment]
Test failed...

**Comment by saeedrahimi** [comment]
>@zombieJ: Test failed...
Done :)

**Comment by DeveloperFromUkraine** [comment]
Hither, what about Tabs ??? Do they support RTL ?

**Comment by zombieJ** [comment]
> Hither, what about Tabs ??? Do they support RTL ?

@saeedrahimi, could we add Tabs in Demo also?

**Comment by DeveloperFromUkraine** [comment]
> > Hither, what about Tabs ??? Do they support RTL ?
> 
> @saeedrahimi, could we add Tabs in Demo also?

Yeah, please if what I talk about this `::after` element of active Tab.

**Comment by saeedrahimi** [comment]
> > Hither, what about Tabs ??? Do they support RTL ?
> 
> @saeedrahimi, could we add Tabs in Demo also?

Tabs is part of [reac-components](https://github.com/react-component/tabs) project and it should supported in that repository.
although i've created a [pull request](https://github.com/react-component/tabs/pull/227) on rc-tabs repository and implemented rtl support on it.

in ant-design some components need to change from their repository to support rtl.

**Review by shakesBeardZ — APPROVED** [review]

**Review by shakesBeardZ — DISMISSED** [review]

**Review by issamYahiaoui — DISMISSED** [review]
Works fine for me

**Review by MRezaSafari — DISMISSED** [review]
Well Done.

**Review by mdeljavan — DISMISSED** [review]

**Comment by shakesBeardZ** [comment]
@zombieJ can you tell us when you plan to merge this branch and ship it?

**Comment by MRezaSafari** [comment]
@zombieJ any idea about this PR? is it gonna merge anytime soon?

File: components/style/core/rtl.less (hunk 11 above)
Comment (saeedrahimi) [reply components/style/core/rtl.less]: @afc163 , I removed `rtl` global styles and added prefixed class to every component. like `ant-cascader-picker-rtl`
Also i fixed grid system direction with another approach.

**Comment by afc163** [comment]
It is better to add this feature in 4.0, could you change target branch to `4.0-prepare`?

**Comment by saeedrahimi** [comment]
> It is better to add this feature in 4.0, could you change target branch to `4.0-prepare`?

target branch changed to  `4.0-prepare`

**Comment by afc163** [comment]
You should check out from `4.0-prepare`.

**Comment by saeedrahimi** [comment]
> You should check out from `4.0-prepare`.

and open a new pull request?

**Comment by afc163** [comment]
> and open a new pull request?

Righto

**Comment by saeedrahimi** [comment]
Moved to a new pull request #19380
